index.js

```diff
@@ -0,0 +1,729 @@
+           equalsY != null && equalsY(x);
+  };
+  Z.equals = equals;
+
+  //# concat :: Semigroup a => a -> a -> a
+  //.
+  //. See [`concat`][concat].
+  //.
+  //. ```javascript
+  //. > concat('abc', 'def')
+  //. 'abcdef'
+  //.
+  //. > concat([1, 2, 3], [4, 5, 6])
+  //. [1, 2, 3, 4, 5, 6]
+  //.
+  //. > concat({x: 1, y: 2}, {y: 3, z: 4})
+  //. {x: 1, y: 3, z: 4}
+  //.
+  //. > concat(Identity('abc'), Identity('def'))
+  //. Identity('abcdef')
+  //. ```
+  var concat = function(x, y) {
+    return Semigroup.getBoundMethod('concat', x)(y);
+  };
+  Z.concat = concat;
+
+  //# empty :: Monoid m => m -> m
+  //.
+  //. See [`empty`][empty].
+  //.
+  //. ```javascript
+  //. > empty('abc')
+  //. ''
+  //.
+  //. > empty([1, 2, 3])
+  //. []
+  //.
+  //. > empty({x: 1, y: 2})
+  //. {}
+  //.
+  //. > empty(Identity([1, 2, 3]))
+  //. Identity([])
+  //. ```
+  var empty = function(x) {
+    return Monoid.getBoundMethod('empty', x)();
+  };
+  Z.empty = empty;
+
+  //# map :: Functor f => (a -> b, f a) -> f b
+  //.
+  //. See [`map`][map].
+  //.
+  //. ```javascript
+  //. > map(x => x + 1, [1, 2, 3])
+  //. [2, 3, 4]
+  //.
+  //. > map(x => x + 1, {x: 1, y: 2})
+  //. {x: 2, y: 3}
+  //.
+  //. > map(x => x + 1, xs => xs.length)('abc')
+  //. 4
+  //.
+  //. > map(x => x + 1, Identity(42))
+  //. Identity(43)
+  //. ```
+  var map = function(f, functor) {
+    return Functor.getBoundMethod('map', functor)(f);
+  };
+  Z.map = map;
+
+  //# bimap :: Bifunctor f => f a c ~> (a -> b, c -> d) -> f b d
+  //.
+  //. See [`bimap`][bimap].
+  var bimap = function(f, g, bifunctor) {
+    return Bifunctor.getBoundMethod('bimap', bifunctor)(f, g);
+  };
+  Z.bimap = bimap;
+
+  //# promap :: Profunctor p => p b c ~> (a -> b, c -> d) -> p a d
+  //.
+  //. See [`promap`][promap].
+  var promap = function(f, g, profunctor) {
+    return Profunctor.getBoundMethod('promap', profunctor)(f, g);
+  };
+  Z.promap = promap;
+
+  //# ap :: Apply f => (f (a -> b), f a) -> f b
+  //.
+  //. See [`ap`][ap].
+  //.
+  //. ```javascript
+  //. > ap([Math.sqrt, x => x * x], [1, 4, 9, 16, 25])
+  //. [1, 2, 3, 4, 5, 1, 16, 81, 256, 625]
+  //.
+  //. > ap(s => n => s.slice(0, n), s => Math.ceil(s.length / 2))('Ramda')
+  //. 'Ram'
+  //.
+  //. > ap(Identity(x => x + 1), Identity(42))
+  //. Identity(43)
+  //. ```
+  var ap = function(applyF, applyX) {
+    return Apply.getBoundMethod('ap', applyX)(applyF);
+  };
+  Z.ap = ap;
+
+  //# chain :: Chain m => ((a -> m b), m a) -> m b
+  //.
+  //. See [`chain`][chain].
+  //.
+  //. ```javascript
+  //. > chain(x => [x, x], [1, 2, 3])
+  //. [1, 1, 2, 2, 3, 3]
+  //. ```
+  var chain = function(f, chain) {
+    return Chain.getBoundMethod('chain', chain)(f);
+  };
+  Z.chain = chain;
+
+  //# chainRec :: ChainRec m => (m x, (a -> c, b -> c, a) -> m c, a) -> m b
+  var chainRec = function(chainRec, f, x) {
+    return ChainRec.getBoundMethod('chainRec', chainRec)(f, x);
+  };
+  Z.chainRec = chainRec;
```

Comment:
@safareli and others, with partial application in mind what is the best argument order for `Z.chainRec`? It's clear to me that `f` should come before `x`, leaving three possibilities:

  - `(chainRec, f, x)`
  - `(f, chainRec, x)`
  - `(f, x, chainRec)`

If `f` uses Array as its ChainRec, `chainRec` must be an array. Thus the third possibility isn't useful: one would never wish to provide `f` without also providing a matching `chainRec`.

There *is* value, though, in providing `chainRec` without `f`. One could write:

```javascript
const chainRecArray = Z.chainRec([]);
```

`(chainRec, f, x)` appears to be the best order. Do you agree? Have I overlooked something important?